cookiejar.go

```diff
@@ -0,0 +1,221 @@
+		t       = time.Now()
+		n       = len(cookies)
+	)
+	for i := 0; i < len(cookies); i++ {
+		c := cookies[i]
+		if !c.Expire().Equal(CookieExpireUnlimited) && c.Expire().Before(t) { // cookie expired
+			cookies = append(cookies[:i], cookies[i+1:]...)
+			ReleaseCookie(c)
+			i--
+		}
+	}
+	// has any cookie been deleted?
+	if n > len(cookies) {
+		cj.hostCookies[hostStr] = cookies
+	}
+	return cookies
+}
+
+// Set sets cookies for a specific host.
+//
+// The host is get from uri.Host().
+//
+// If the cookie key already exists it will be replaced by the new cookie value.
+func (cj *CookieJar) Set(uri *URI, cookies ...*Cookie) {
+	if uri != nil {
+		cj.set(uri.Host(), cookies...)
+	}
+}
+
+// SetByHost sets cookies for a specific host.
+//
+// If the cookie key already exists it will be replaced by the new cookie value.
+func (cj *CookieJar) SetByHost(host []byte, cookies ...*Cookie) {
+	cj.set(host, cookies...)
+}
+
+func (cj *CookieJar) set(host []byte, cookies ...*Cookie) {
+	cj.init()
+
+	cj.m.Lock()
+	{
+		hostStr := b2s(host)
+		hcs, ok := cj.hostCookies[hostStr]
+		if !ok {
+			// If the key does not exists in the map then
+			// we must make a copy for the key.
+			hostStr = string(host)
+		}
+		for _, cookie := range cookies {
+			c := searchCookieByKeyAndPath(cookie.Key(), cookie.Path(), hcs)
+			if c != nil {
+				c.ParseBytes(cookie.Cookie())
+			} else {
+				c = AcquireCookie()
+				c.CopyTo(cookie)
+				hcs = append(hcs, c)
+			}
+		}
+		cj.hostCookies[hostStr] = hcs
+	}
+	cj.m.Unlock()
+}
+
+// SetKeyValue sets a cookie by key and value for a specific host.
+//
+// This function prevents extra allocations by making repeated cookies
+// not being duplicated.
+func (cj *CookieJar) SetKeyValue(host, key, value string) {
+	cj.SetKeyValueBytes(host, s2b(key), s2b(value))
+}
+
+// SetKeyValueBytes sets a cookie by key and value for a specific host.
+//
+// This function prevents extra allocations by making repeated cookies
+// not being duplicated.
+func (cj *CookieJar) SetKeyValueBytes(host string, key, value []byte) {
+	cj.setKeyValue(host, key, value)
+}
+
+func (cj *CookieJar) setKeyValue(host string, key, value []byte) {
+	c := AcquireCookie()
+	c.SetKeyBytes(key)
+	c.SetValueBytes(value)
+	cj.set(s2b(host), c)
+}
+
+func (cj *CookieJar) dumpTo(req *Request) {
+	uri := req.URI()
+	cookies := cj.get(uri.Host(), uri.Path())
+	for _, cookie := range cookies {
+		req.Header.SetCookieBytesKV(cookie.Key(), cookie.Value())
+	}
+}
+
+func (cj *CookieJar) getFrom(host, path []byte, res *Response) {
+	cj.init()
+
+	hostStr := b2s(host)
+	cj.m.Lock()
+	{
+		cookies, ok := cj.hostCookies[hostStr]
+		if !ok {
+			// If the key does not exists in the map then
+			// we must make a copy for the key.
+			hostStr = string(host)
+		}
+		t := time.Now()
+		res.Header.VisitAllCookie(func(key, value []byte) {
+			created := false
+			c := searchCookieByKeyAndPath(key, path, cookies)
+			if c == nil {
+				c = AcquireCookie()
+				created = true
```

Comment:
Better do the expired check here first so you don't acquire and release a cookie immediately if it didn't exist already.

Reply:
I don't see any other way to do this.
If I do the expiration checking at cookiejar.go:197 first I need to call ParseBytes and the code will be repeated in the case `c == nil` and `c != nil`. I prefer to keep it like this unless you see any other way to do that.

Reply:
Yes you are right, I forgot that you have to somehow parse the cookie first to get the expiration date. Doing that in a stack allocate cookie value would be possible for example but it would result in messier code.